docs/proposals/loadbalancing.md

```diff
@@ -0,0 +1,338 @@
+  * Detecting when a single loadbalancer is saturated can be hard
+
+### Issues
+
+What follows is a digest of some of the problems brought up in various github issues. Expect these to make it into the proposal above in due time.
+
+#### How does this work at L4?
+
+People choose L4 for a few different reasons:
+- Cost
+- Latency
+- Throughput of loadbalancer is the bottleneck
+- Cloud provider limitations
+- Want SSL passthrough or SNI switch for multi-certs
+- Need client IP and cannot set loadbalancer proxy as default gateway
+
+Point being, we can't reliably infer all these use case with Service.Type=LoadBalancer or the absence of a UrlMap in the IngressPoint (because they might want L7 for session persistence). Instead, anything serving ingress traffic needs an ingress path (which is why it isn't called urlMap). An L4 ingress path might look like:
+
+```yaml
+apiVersion: v1
+kind: IngressPoint
+metadata:
+  name: l4ingress
+  type: network
+spec:
+  host: www.example.com
+  tlsMode: Termination
+  secret: foosecret
+status:
+  loadBalancer:
+    ingress:
+    - ip: 104.1..
+```
+
+type: network with a pathmap is a validation error.
+This adds a slight burden on users, since what was previously a one line change to expose a service now requires a new resource.
+
+
+### Why not use a single loadbalancer for all Ingress paths?
+
+Use cases for not having all ingress paths join an existing load balancer:
+* User wants only foo.com and bar.com backed by a single ip
+* Internal services get bare metal loadbalancing, user facing services are backed by a cloud loadbalancer
+* User already has a specific ip in DNS, and wants to expose a loadbalanced website on it
+* A user wants to isolate their host from decryption of other hosts saturating cpu on a single bare metal loadbalancer (this isn't really a problem for cloud) by explicitly asking for a new one.)
+
+### How does the ingress path choose a loadbalancer to join?
+
+For 1.1, there will only be one loadbalancer controller that claims all ingress paths. Whether it creates a new loadbalancer or reuses an existing one is upto the controller. For example, the GCE controller will create 1 loadbalancer per ingress path, if it has `type: network`, this will be an l4, if it has `type: application` it will be an l7 with rules to allow proxying to different backends based on the given pathmap.
+
+However we need to provide a way for users to specify which class of loadbalancers to join, and which particular loadbalancer (in the case of multiple loadbalancer per class). Note that we can't just say each class of loadbalancers has a single loadbalancer in it, and so picking a class boils down to picking a specific loadbalancer, because of cloud provider limitations. Nor can we say picking a specific loadbalancer picks the class, because a loadbalancer might have limited capacity (i.e don't force a user to pick 172.12, which might be maxed out, when they just want class=gce).
```

Comment:
An alternative may be to allow users can provide some hints as to capacity requirements [ala ```bandwidth```, ```concurrent-sessions```, ```SLA-level``` as an example] with the ingress path and the system could ```best fit``` the class/LB to use - could do this via annotations as well potentially. Note, this is not going to solve all cases as it could be a case of a decision made at a point-in-time when say the balancer had capacity and the reality at some future time might be the cup runneth over!! But that could well be addressed by increasing capacity.